crates/rethnet_eth/src/remote/client.rs

```diff
@@ -269,9 +273,100 @@ impl RpcClient {
         }
     }
 
+    /// Return the largest known block number if it's considered safe for the given block, otherwise
+    /// fetch the latest block number.
+    async fn largest_known_or_latest_block_number(
+        &self,
+        safety_checker: &CacheKeyForUncheckedBlockNumber<'_>,
+        chain_id: &U256,
+    ) -> Result<U256, RpcClientError> {
+        let largest_known_block_number = { *self.largest_known_block_number.read().await };
+        if let Some(largest_known_block_number) = largest_known_block_number {
```

Comment:
It’s ok if the largest known block number is stale, because we only use it to check if a certain block number is safe to cache relative to it. And if we determine that it’s unsafe, then we fetch the latest block number and check again. So basically it lets us potentially save a request.

Reply:
Refactored the code to make it clearer: [17dd0da](https://github.com/NomicFoundation/hardhat/pull/4319/commits/17dd0da925174f0d648ca40436da462f63012a03)